File: src/byte/orchestration/state.py
from typing import Annotated, Dict, TypedDict, Union

from langgraph.graph.message import AnyMessage, add_messages

from byte.orchestration import ConstraintSchema, MetadataSchema, PhaseModel, Reducer, RoutePhaseModel


class HarnessState(TypedDict):
    """State passed to a agents, constraining its skills, xyz."""

    spec: str
    skills: list[str]


class RoutingState(TypedDict):
    """Routing information for graph node transitions."""

    target: str
    source: str


class BaseState(TypedDict):
    """Base state that all agents inherit with messaging and status tracking."""

    harness: HarnessState

    # Persistent conversation history from memory store
    history_messages: Annotated[list[AnyMessage], add_messages]

    # Ephemeral messages for current execution only (validation, errors, etc.)
    scratch_messages: Annotated[list[AnyMessage], add_messages]

    final_message: AnyMessage | None

    # Current user request being processed by the agent
    user_request: str

    constraints: Annotated[list[ConstraintSchema], Reducer.add_constraints]

    errors: Annotated[str | None, Reducer.replace_str]

    # These are specific to Coder
    touched_files: list[str]

    workflow_phases: Dict[str, Union[PhaseModel | RoutePhaseModel]] | None

    routing: RoutingState

    is_cancelled: bool

    metadata: Annotated[MetadataSchema, Reducer.update_metadata]

File: node/src/graphs/state.ts
import { StateSchema, MessagesValue } from "@langchain/langgraph";
import { BaseMessage } from "@langchain/core/messages";
import { z } from "zod";
import {
  MetadataSchema,
  ConstraintSchema,
  type Metadata,
  type Constraint,
} from "./schemas.ts";

// Custom reducers for state management
const addConstraints = (
  current: Constraint[],
  update: Constraint | Constraint[],
): Constraint[] => {
  const updates = Array.isArray(update) ? update : [update];
  return [...current, ...updates];
};

const replaceStr = (
  current: string | null,
  update: string | null,
): string | null => {
  return update;
};

const updateMetadata = (
  current: Metadata,
  update: Partial<Metadata>,
): Metadata => {
  return { ...current, ...update };
};

// Base state schema that all agents inherit
export const BaseStateSchema = new StateSchema({
  // Persistent conversation history from memory store
  history_messages: MessagesValue,

  // Ephemeral messages for current execution only (validation, errors, etc.)
  scratch_messages: MessagesValue,

  // Current user request being processed by the agent
  user_request: z.string(),

  constraints: z.array(ConstraintSchema).default(() => []),

  masked_messages: z.array(z.custom<BaseMessage>()).default(() => []),

  agent: z.string(),

  errors: z.string().nullable().default(null),

  examples: z.array(z.custom<BaseMessage>()).default(() => []),

  extracted_content: z
    .union([z.string(), z.record(z.unknown())])
    .nullable()
    .default(null),

  // These are specific to Coder
  parsed_blocks: z.array(z.record(z.unknown())).default(() => []),

  // This is specific to subprocess
  command: z.string().default(""),

  metadata: MetadataSchema.default({}),
});

// Export the state type
export type BaseState = typeof BaseStateSchema.State;
